letsencrypt/client/proof_of_possession.py

```diff
@@ -0,0 +1,82 @@
+"""Proof of Possession Identifier Validation Challenge."""
+import M2Crypto
+import os
+import zope.component
+
+from letsencrypt.acme import challenges
+from letsencrypt.acme import jose
+from letsencrypt.acme import other
+from letsencrypt.client import interfaces
+from letsencrypt.client.display import util as display_util
+
+
+class ProofOfPossession(object): # pylint: disable=too-few-public-methods
+    """Proof of Possession Identifier Validation Challenge.
+
+    Based on draft-barnes-acme, section 6.5.
+
+    :ivar installer: Installer object
+    :type installer: :class:`~letsencrypt.client.interfaces.IInstaller`
+
+    """
+    def __init__(self, installer):
+        self.installer = installer
+
+    def perform(self, achall):
+        """Perform the Proof of Possession Challenge.
+
+        :param achall: Proof of Possession Challenge
+        :type achall: :class:`letsencrypt.client.achallenges.ProofOfPossession`
+
+        :returns: Response or None/False if the challenge cannot be completed
+        :rtype: :class:`letsencrypt.acme.challenges.ProofOfPossessionResponse`
+            or False
+
+        """
+        if (not isinstance(achall.challb.hints.jwk, achall.challb.alg.kty) or
+                achall.challb.alg in [jose.HS256, jose.HS384, jose.HS512]):
+            return None
+
+        # This will work regardless of how JWKES is implemented
+        for cert, key, _ in self.installer.get_all_certs_keys():
+            der_cert_key = M2Crypto.X509.load_cert(cert).get_pubkey().as_der()
+            cert_key = achall.challb.alg.kty.load(der_cert_key)
```

Comment:
I believe you should be able to proxy directly from achall.  So achall.alg.kty.load should work.